### ISSUE TITLE
How to run on Mac?

### ISSUE THREAD

**Issue by thomasd3** [Thraka/SadConsole#138]
From my understanding the SDL2 package for mac needs to be installed, but while the console lib is compiling fine, the example just will not run.

Is it possible to get a quick write up about how to build / run on MacOS?

**Comment by thomasd3** [comment]
I tried a few things:

- Get the latest and open the solution with mac only in it. This will fail trying to load the SDL2 dll.
- Create an empty project and bring in the nuget sadconsole package, similar result
- Create an empty project and bring in the nuger sadconsolefna package and I can't compile the init code from http://sadconsole.com/docs/monogame-template.html because monogame and sadconsole's version of the xna port libs seem to differ.

**Comment by Chris3606** [comment]
I haven't tried it on Mac, however I have linux.  It _should_ (operative word should) be a matter of:
1. Get newest SadConsole.Starter and MonoGame.Desktop.OpenGL from NuGet as per your first attempt.
2. Copy the `libopenal.1.dylib` and `libSDL-2.0.0.dylib` files from the MonoGame project template to the project directory, and make sure they're copied to build directory on project build.
3. Make sure the `MonoGame.Framework.dll.config` file has at least the following 2 entries in the `<configuration>` element:
```
<dllmap dll="SDL2.dll" os="osx" target="libSDL2-2.0.0.dylib"/>
<dllmap dll="soft_oal.dll" os="osx" target="libopenal.1.dylib" />
```
This should get it to use the mac binaries as its SDL2 library.  The process is similar on Linux, just with the Linux library files.  This is also very similar to the way the MonoGame template sets things up.

A lot of this is guesswork/from memory, and I haven't ever tried it on Mac, so I can't guarantee much -- I'm sorry I can't be more helpful -- hopefully somebody else knows more.

**Comment by thomasd3** [comment]
I managed to make it work:

I created a .net standard lib and I imported the source of SadConsole and I imported the MonoGame.Framework.DesktopGL.Core package from nuget.
(btw, some files in the SadConsole source tree should NOT be included)

When I used the existing packages on Nuget, no matter the combinations I tried, I couldn't get it to work.

**Comment by thomasd3** [comment]
Yes, definitely; once I included it along with the source of SadConsole, it worked right away without a single tweak.

I’m assuming I’m not the only one having had issues on the mac, so it could definitely be useful.

And, from my perspective, instead of including your source, it makes a lot more sense to use your package as well.

Having starting in the 8bit era, I always find frustrating that getting something on screen takes so much work nowadays; thanks to your work, this is becoming simpler!

Thomas.

**Comment by Thraka** [comment]
That is a big thing I'm focusing on with V7 that is in the pipeline right now. Simplifying terminology and how things work. Please capture some feedback and give it to me! I also realize I have a lot of gaps in the documentation. When V7 is completed, I'm going to do a property "basics" tutorial series that does't necessarily focus on make a roguelike, but instead just focuses on exploring parts of SadConsole.

**Comment by thomasd3** [comment]
Is V7 in the develop branch? Or, is it advanced enough that it can be used at this stage?

**Comment by thomasd3** [comment]
btw, I noticed something: you have the folder themes, but in the folder serialized types, you have a copy of the themes files again.

If the V7 branch is advanced enough, I can verify that it works on Mac.

Have you thought about detaching the rendering and update/display calls completely from XNA with an interface? For example, on Mac doing the rendering with Cocos2d woudl be significantly simpler

**Comment by Thraka** [comment]
That is a difference between how Standard library projects work and .NET Framework libraries work. Something I didn't know about. The Standard libraries (or Core projects) just use all files under the folder. That extra serialization stuff and some of the stuff in `gamehelpers` is not actually included in SadConsole because they were removed or not finished. I was a bit messy with my Master branch. Bad on me!

I have thought about decoupling from MonoGame, but the problem with it is that there are so many primitives used from MonoGame, that it's hard to translate things all in a single codebase. For example, I actually supported the SFML game engine at one point but the differences between their `Point` and `Rectangle` classes and other stuff meant that I had to put in tons of `#if` statements all over the place. Colors were totally handled differently and a lot of static methods provided by MonoGame didn't exist.

**Comment by thomasd3** [comment]
We had similar issues with several products since we did a lot of cross platform ports (> 25 years in video games) where you’d have engines built for one tech that had to be ported to other platforms, etc.

The only viable long term strategy with this issue is to define your primitives based on what your code is already using (XNA in this case, which is good since it's well designed) and force each person writing a driver for a different platform to handle the translation at their level.

Ultimately, ignoring kb/mouse, you probably just need a framework of callbacks (init, update, draw, etc), math and geometry primitives, a standard way to handle colors, textures and buffers with the ability to draw quads and perform matrix operations.
The API to communicate to the renderer is probably not that large in the end.

Since you’re bound to XNA/MonoGame, the easiest is to copy their objects and definitions to make them your own and the XNA driver would become a very simple piece of code.
You’d be essentially passing the bulk of the work to people writing the drivers, but that would open quite a few doors.

**Comment by Thraka** [comment]
Thanks for the tips. You can kind of see that I was thinking about that by putting the MonoGame Game stuff in the [Graphics.MonoGame](https://github.com/Thraka/SadConsole/tree/master/src/SadConsole.Shared/Graphics.MonoGame) namespace 😁 

I didn't catch your other question. The V7 is a different develop branch (develop-v7 i think) and it's advanced enough but I'm in a middle of a large commit that will restructure a lot of surface/rendering/console/screen object model into something simpler and easier to use. Also renaming a lot of stuff to make more sense as to intent, and I'll also be working on redoing the controls system a bit.

**Comment by thomasd3** [comment]
yes, I saw the Graphics.MonoGame namespace!

As soon you’re done with the v7 commit with the new API, I’ll make sure to port what I’m working on to the new API and test things on Mac.
